File: python-fastapi/app/schemas.py
from datetime import datetime
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from .models import CrewmateRole, MissionStatus, ShipClass


class ErrorDetail(BaseModel):
    field: str
    message: str


class ErrorResponse(BaseModel):
    error: str
    details: Optional[list[ErrorDetail]] = None


class ShipRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    name: str
    ship_class: ShipClass = Field(serialization_alias="class")
    registry: str
    warp_capable: bool
    created_at: datetime
    updated_at: datetime


class ShipCreate(BaseModel):
    name: str
    ship_class: ShipClass = Field(alias="class")
    registry: str
    warp_capable: bool = False

    model_config = ConfigDict(populate_by_name=True)


class ShipUpdate(BaseModel):
    name: Optional[str] = None
    ship_class: Optional[ShipClass] = Field(default=None, alias="class")
    registry: Optional[str] = None
    warp_capable: Optional[bool] = None

    model_config = ConfigDict(populate_by_name=True)


class CrewmateRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    ship_id: UUID
    name: str
    role: CrewmateRole
    species: str
    crew_rank: int = Field(serialization_alias="rank")
    created_at: datetime
    updated_at: datetime


class CrewmateCreate(BaseModel):
    ship_id: UUID
    name: str
    role: CrewmateRole
    species: str
    crew_rank: int = Field(alias="rank")

    model_config = ConfigDict(populate_by_name=True)


class CrewmateUpdate(BaseModel):
    ship_id: Optional[UUID] = None
    name: Optional[str] = None
    role: Optional[CrewmateRole] = None
    species: Optional[str] = None
    crew_rank: Optional[int] = Field(default=None, alias="rank")

    model_config = ConfigDict(populate_by_name=True)


class MissionRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    ship_id: UUID
    codename: str
    objective: str
    status: MissionStatus
    started_at: Optional[datetime]
    ended_at: Optional[datetime]
    created_at: datetime
    updated_at: datetime


class MissionCreate(BaseModel):
    ship_id: UUID
    codename: str
    objective: str
    status: MissionStatus = MissionStatus.planned
    started_at: Optional[datetime] = None
    ended_at: Optional[datetime] = None


class MissionUpdate(BaseModel):
    ship_id: Optional[UUID] = None
    codename: Optional[str] = None
    objective: Optional[str] = None
    status: Optional[MissionStatus] = None
    started_at: Optional[datetime] = None
    ended_at: Optional[datetime] = None

File: node-express/src/schemas.js
import { z } from "zod";

export const shipClassEnum = z.enum(["cruiser", "frigate", "destroyer", "scout"]);
export const crewmateRoleEnum = z.enum(["captain", "engineer", "medic", "pilot", "gunner"]);
export const missionStatusEnum = z.enum(["planned", "active", "completed", "failed"]);

export const shipCreateSchema = z.object({
  name: z.string().min(1),
  class: shipClassEnum,
  registry: z.string().min(1),
  warp_capable: z.boolean().optional().default(false),
});

export const shipUpdateSchema = z
  .object({
    name: z.string().min(1).optional(),
    class: shipClassEnum.optional(),
    registry: z.string().min(1).optional(),
    warp_capable: z.boolean().optional(),
  })
  .refine((data) => Object.keys(data).length > 0, {
    message: "At least one field is required",
  });

export const crewmateCreateSchema = z.object({
  ship_id: z.string().uuid(),
  name: z.string().min(1),
  role: crewmateRoleEnum,
  species: z.string().min(1),
  rank: z.number().int(),
});

export const crewmateUpdateSchema = z
  .object({
    ship_id: z.string().uuid().optional(),
    name: z.string().min(1).optional(),
    role: crewmateRoleEnum.optional(),
    species: z.string().min(1).optional(),
    rank: z.number().int().optional(),
  })
  .refine((data) => Object.keys(data).length > 0, {
    message: "At least one field is required",
  });

export const missionCreateSchema = z.object({
  ship_id: z.string().uuid(),
  codename: z.string().min(1),
  objective: z.string().min(1),
  status: missionStatusEnum.optional().default("planned"),
  started_at: z.string().datetime().nullable().optional(),
  ended_at: z.string().datetime().nullable().optional(),
});

export const missionUpdateSchema = z
  .object({
    ship_id: z.string().uuid().optional(),
    codename: z.string().min(1).optional(),
    objective: z.string().min(1).optional(),
    status: missionStatusEnum.optional(),
    started_at: z.string().datetime().nullable().optional(),
    ended_at: z.string().datetime().nullable().optional(),
  })
  .refine((data) => Object.keys(data).length > 0, {
    message: "At least one field is required",
  });

export function formatZodError(error) {
  return {
    error: "validation_failed",
    details: error.errors.map((e) => ({
      field: e.path.join(".") || "body",
      message: e.message,
    })),
  };
}
